include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_merge.h

```diff
@@ -166,34 +272,394 @@ struct __parallel_merge_submitter<_IdType, __internal::__optional_kernel_name<_N
+                [=](sycl::nd_item</*dim=*/1> __nd_item)
+                {
+                    const std::size_t __global_linear_id = __nd_item.get_global_linear_id();    // Merge matrix diagonal's GLOBAL index
+                    const std::size_t __local_id = __nd_item.get_local_id(0);                   // Merge sub-matrix LOCAL diagonal's index
+                    const std::size_t __group_linear_id = __nd_item.get_group_linear_id();      // Merge matrix base diagonal's GLOBAL index
+
+                    _IdType __rng1_wg_data_size = 0;
+                    _IdType __rng2_wg_data_size = 0;
+
+                    _Range1ValueType* __rng1_cache_slm = nullptr;
+                    _Range1ValueType* __rng2_cache_slm = nullptr;
+
+                    // Current diagonal inside of the merge matrix?
+                    const bool __have_data = __global_linear_id * __chunk < __n;
+
+                    // Current diagonal inside of the merge matrix?
+                    if (__have_data)
+                    {
+                        // Split points on left anr right base diagonals
+                        //  - in GLOBAL coordinates
+                        const _split_point_t<std::size_t>& __sp_base_left_global  = __base_diagonals_sp_global_ptr[__group_linear_id];
+                        const _split_point_t<std::size_t>& __sp_base_right_global = __base_diagonals_sp_global_ptr[__group_linear_id + 1]; 
+
+                        assert(__sp_base_right_global.first >= __sp_base_left_global.first);
+                        assert(__sp_base_right_global.second >= __sp_base_left_global.second);
+
+                        __rng1_wg_data_size = __sp_base_right_global.first - __sp_base_left_global.first;
+                        __rng2_wg_data_size = __sp_base_right_global.second - __sp_base_left_global.second;
+
+                        auto [__rng1_loc_acc, __offset_to_slm1] = __merge_slm_helper::template get_local_accessor<0>(__loc_acc_pack);
+                        auto [__rng2_loc_acc, __offset_to_slm2] = __merge_slm_helper::template get_local_accessor<1>(__loc_acc_pack, __rng1_wg_data_size);
+                        __rng1_cache_slm = std::addressof(__rng1_loc_acc[0]) + __offset_to_slm1;
+                        __rng2_cache_slm = std::addressof(__rng2_loc_acc[0]) + __offset_to_slm2;
+
+                        // Cooperative data load from __rng1 to __rng1_cache_slm, from __rng2 to __rng1_cache_slm
+                        load_data_into_slm(__rng1, __rng1_cache_slm, __sp_base_left_global.first,  __sp_base_right_global.first,
+                                           __rng2, __rng2_cache_slm, __sp_base_left_global.second, __sp_base_right_global.second,
+                                           __wi_in_one_wg, __local_id);
+                    }
+
+                    // Wait until all the data is loaded (if we have more then one item in work-group
+                    if (__wi_in_one_wg > 1)
```

Comment:
This is not the issue, but it seems that having a branch here for this degenerate case is probably best avoided at a higher level, or not worth caring about.